Fix inconsistencies and broken key bindings in the Rust layer

bcc32 opened the pull request:
The Rust layer was recently overhauled in #16078, but unfortunately the large rework caused some breaking changes.

This PR attempts to fix all the breakages I've noticed (although there may be others).

Unfortunately, Rustic does not support "repeating" the last cargo command, and also it is much more inconvenient to try to "auto-reload" rust-analyzer because Rustic runs all cargo commands asynchronously, so I have dropped that functionality (which was not working correctly anyhow).

(If folks would like me to attempt to resurrect that feature, I can try to do so, but most likely the best way to do that would be a patch to Rustic itself.)

bcc32 committed:
[rust] Resurrect rust-backend variable

rust-backend was removed but is still meaningful (needs to be set to
the symbol lsp in order for LSP support to be enabled).

This commit restores the variable to config.el and removes boundp
checks because it will always be bound now (except in layers.el which
is loaded before config.el).

bcc32 committed:
[rust] Fully remove support for unmaintained rls backend

bcc32 committed:
[rust] Bind SPC m c . to rustic-cargo-run-rerun

This command has different behavior than cargo-process-repeat (the
previous meaning of this binding), but it is still a reasonable
binding and is the most similar command that exists in the rustic
package.

bcc32 committed:
[rust] Fix wrong command name

bcc32 committed:
[rust] Remove wrapper commands for package cargo

This package was removed in the Rust layer overhaul, so these commands
are not used anywhere.  In fact, the reloading behavior cannot easily
be implemented with rustic, because rustic runs `cargo add` and the
like in a compilation-mode buffer rather than synchronously.

As a result, we also remove references to automatically reloading the
workspace in rust-analyzer and the corresponding config variable.

bcc32 committed:
[rust] Remove unnecessary post-init stage

Layer rust owns package rustic, so there is no need to use post-init.

bcc32 committed:
[rust] Ensure rustic-mode key bindings are always available

Using :after in use-package prevents the :init forms from being
evaluated until lsp-mode and flycheck are loaded.  However, if the
user has disabled lsp-mode or flycheck, they should still be able to
use the rustic-cargo-* key bindings, so :after is not appropriate
here.

Instead, directly use with-eval-after-load where necessary for the
lsp-mode- and flycheck-specific configuration.

bcc32 committed:
[rust] Add rustic-clippy as an available flycheck checker

This just uncomments some code that was already there.  Not sure why
it was commented in the first place.

bcc32 committed:
[rust] Remove redundant progn

whirm commented:
There's `rustic-cargo-run-rerun`, I'm aware it doesn't do exactly the same but maybe we can bind that in its place?

`lsp-rust-analyzer-(re)run` is what I personally use and I'm not sure I see the value of repeating the last cargo command other than for `cargo run` + args, which is already implemented in rustic.

Regarding auto reloading rust-analyzer. I'd say it just works by default. If you modify `Cargo.toml`  outside of emacs (IE by calling `cargo add foobar` or by modifying it with a different editor) RA will notice and reload by itself. Just double checked on my machine.

FWIW, I've reviewed and I'm running your changes locally and everything looks fine to me.

Thanks @bcc32 and @smile13241324!

bcc32 commented:
Thanks @whirm.  I've incorporated your suggestion w.r.t. `,c.` and also reworked the comment about auto-reloading the workspace to reflect the current state of the world.
